--- Day 7: Bridge Repair ---
The Historians take you to a familiar rope bridge over a river in the middle of a jungle. The Chief isn't on this side of the bridge, though; maybe he's on the other side?

When you go to cross the bridge, you notice a group of engineers trying to repair it. (Apparently, it breaks pretty frequently.) You won't be able to cross until it's fixed.

You ask how long it'll take; the engineers tell you that it only needs final calibrations, but some young elephants were playing nearby and stole all the operators from their calibration equations! They could finish the calibrations if only someone could determine which test values could possibly be produced by placing any combination of operators into their calibration equations (your puzzle input).

For example:

190: 10 19
3267: 81 40 27
83: 17 5
156: 15 6
7290: 6 8 6 15
161011: 16 10 13
192: 17 8 14
21037: 9 7 18 13
292: 11 6 16 20
Each line represents a single equation. The test value appears before the colon on each line; it is your job to determine whether the remaining numbers can be combined with operators to produce the test value.

Operators are always evaluated left-to-right, not according to precedence rules. Furthermore, numbers in the equations cannot be rearranged. Glancing into the jungle, you can see elephants holding two different types of operators: add (+) and multiply (*).

Only three of the above equations can be made true by inserting operators:

190: 10 19 has only one position that accepts an operator: between 10 and 19. Choosing + would give 29, but choosing * would give the test value (10 * 19 = 190).
3267: 81 40 27 has two positions for operators. Of the four possible configurations of the operators, two cause the right side to match the test value: 81 + 40 * 27 and 81 * 40 + 27 both equal 3267 (when evaluated left-to-right)!
292: 11 6 16 20 can be solved in exactly one way: 11 + 6 * 16 + 20.
The engineers just need the total calibration result, which is the sum of the test values from just the equations that could possibly be true. In the above example, the sum of the test values for the three equations listed above is 3749.

Determine which equations could possibly be true. What is their total calibration result?



In [46]:
with open('input7.txt') as file:
    data = []
    for line in file:
        solution, problem = line.strip().split(':')
        solution = int(solution)
        problem = [int(num) for num in problem.split()]
        data.append((problem, solution))

data

[([10, 1, 489, 35, 44, 14, 16, 4], 5055314034),
 ([18, 1, 121, 2, 6, 38], 108773),
 ([1, 73, 45, 2, 35, 5], 22175),
 ([338, 1, 6, 1, 191, 7, 3, 4, 34, 4], 4076637638),
 ([9, 7, 73, 1, 6, 4, 7, 8, 581, 7, 2, 8], 479231839),
 ([761, 852, 11, 4, 8, 28, 6], 161312232),
 ([220, 40, 7, 5, 878, 25], 6760600000),
 ([608, 91, 53, 5, 5, 53, 649], 601121975),
 ([79, 2, 6, 558, 42], 761),
 ([65, 90, 8, 6, 9], 84130),
 ([7, 3, 9, 3, 8, 543, 1, 673, 8, 4], 52590888326),
 ([9, 80, 3, 1], 21603),
 ([9, 6, 7, 32, 210], 6498240),
 ([8, 413, 85, 167, 5, 81], 2420318475),
 ([885, 2, 2, 599, 5, 1, 6, 6, 6, 5, 9], 6368572316),
 ([870, 8, 1, 308, 581, 5, 34, 9], 23428423),
 ([1, 8, 50, 1, 76, 67, 395], 17387502),
 ([9, 8, 2, 6, 2, 282, 5, 507, 954], 16073047995028),
 ([2, 8, 95, 6, 8, 8, 152, 6, 1, 9, 5, 8], 1527631),
 ([32, 593, 2, 1, 4, 1, 6, 4, 6, 45, 7], 90086454),
 ([8, 60, 6, 7, 860, 59, 7, 14], 516078606614),
 ([2, 86, 847, 269, 3, 6], 145962),
 ([9, 8, 7, 3, 386, 37, 82, 2, 8], 81188),
 ([50, 1, 764,

In [49]:
def compute(numbers):
    sum_ = numbers[0] + numbers[1]
    product = numbers[0] * numbers[1]
    if len(numbers) == 2:
        return set((sum_, product))

    return compute([sum_] + numbers[2:]) | compute([product] + numbers[2:])


In [51]:
total_calibration_result = 0
for problem, solution in data:
    if solution in compute(problem):
        total_calibration_result += solution
total_calibration_result

932137732557

Your puzzle answer was 932137732557.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The engineers seem concerned; the total calibration result you gave them is nowhere close to being within safety tolerances. Just then, you spot your mistake: some well-hidden elephants are holding a third type of operator.

The concatenation operator (||) combines the digits from its left and right inputs into a single number. For example, 12 || 345 would become 12345. All operators are still evaluated left-to-right.

Now, apart from the three equations that could be made true using only addition and multiplication, the above example has three more equations that can be made true by inserting operators:

156: 15 6 can be made true through a single concatenation: 15 || 6 = 156.
7290: 6 8 6 15 can be made true using 6 * 8 || 6 * 15.
192: 17 8 14 can be made true using 17 || 8 + 14.
Adding up all six test values (the three that could be made before using only + and * plus the new three that can now be made by also using ||) produces the new total calibration result of 11387.

Using your new knowledge of elephant hiding spots, determine which equations could possibly be true. What is their total calibration result?

In [58]:
def compute(numbers):
    sum_ = numbers[0] + numbers[1]
    product = numbers[0] * numbers[1]
    concat = int(str(numbers[0]) + str(numbers[1]))
    if len(numbers) == 2:
        return set((sum_, product, concat))

    return compute([sum_] + numbers[2:]) | compute([product] + numbers[2:]) | compute([concat]+ numbers[2:])


In [59]:
total_calibration_result = 0
for problem, solution in data:
    if solution in compute(problem):
        total_calibration_result += solution
total_calibration_result

661823605105500

That's the right answer! You are one gold star closer to finding the Chief Historian.

You have completed Day 7! You can [Share] this victory or [Return to Your Advent Calendar].